rdrf/rdrf/services/io/notifications/longitudinal_followups.py

```diff
@@ -0,0 +1,163 @@
+    )
+
+    return {longitudinal_followup: list(entries) for longitudinal_followup, entries in grouped}
+
+
+def with_now(func):
+    def wrapper(now=None):
+        return func(now=now or datetime.datetime.now())
+
+    return wrapper
+
+
+@with_now
+def send_longitudinal_followups(now):
+    allowed_registries = [
+        r.code for r in Registry.objects.all() if r.has_feature(RegistryFeatures.LONGITUDINAL_FOLLOWUPS)
+    ]
+
+    outstanding_entries = LongitudinalFollowupEntry.objects.annotate(
+        debounce_value=Coalesce(
+            "longitudinal_followup__debounce",
+            Value(datetime.timedelta(seconds=0))
+        ),
+        send_at_debounced=ExpressionWrapper(
+            F("send_at") - F("debounce_value"),
+            output_field=DateTimeField()
+        ),
+        first_form_name=Subquery(
+            ContextFormGroupItem.objects.filter(
+                context_form_group=OuterRef("longitudinal_followup__context_form_group__id")
+            ).values("registry_form__name")[:1]
+        ),
+        longitudinal_followup_name=F("longitudinal_followup__name"),
+        longitudinal_followup_description=F("longitudinal_followup__description"),
+        context_form_group_name=F("longitudinal_followup__context_form_group__name"),
+        context_form_group_context_type=F("longitudinal_followup__context_form_group__context_type"),
+        registry_code=F("longitudinal_followup__context_form_group__registry__code"),
+    ).filter(
+        registry_code__in=allowed_registries,
+        send_at_debounced__lte=now,
+        state=LongitudinalFollowupQueueState.PENDING,
+        patient__id__in=LongitudinalFollowupEntry.objects.filter(
+            state=LongitudinalFollowupQueueState.PENDING,
+            send_at__lte=now
+        ).values("patient__id").distinct(),
+    ).order_by("patient__id", "created_at")
+
+    logger.info(f"Found {len(outstanding_entries)} outstanding followup entries")
+
+    sent_success = sent_failure = 0
+    for patient_id, patient_entries_group in itertools.groupby(outstanding_entries, lambda entry: entry.patient.id):
+        all_patient_entries = list(patient_entries_group)
+        patient_entries = list(filter(evaluate_condition, all_patient_entries))
+        logger.debug(f"Patient {patient_id}: {len(patient_entries)} / {len(all_patient_entries)} followup entries")
+
+        if len(patient_entries) == 0:
+            continue
+
+        patient = patient_entries[0].patient
+
+        # At least one email that's eligible before debounce
+        assert any(entry.send_at <= now for entry in patient_entries)
+
+        sent_at = datetime.datetime.now()
+
+        longitudinal_followups = serialize_entries(patient_entries)
+
+        for entry in patient_entries:
+            entry.sent_at.append(sent_at)
+            entry.state = LongitudinalFollowupQueueState.SENT
+        LongitudinalFollowupEntry.objects.bulk_update(patient_entries, ["sent_at", "state"])
```

Comment:
I believe we discussed this in a dev meeting, considering the two failure modes:

1. An email is successfully sent to a user's address, but there's an error on our end recording that success
2. An email fails to send to a user's address, and we record it as being sent when it hasn't

The suggestion was to favor the latter, because the former could lead to repeatedly sending emails unwanted emails to a user.

Initially, I had a third `LongitudinalFollowupQueueState` called `FAILED` which would allow admins to handle send failutes, but I believe it was suggested to just use two queue states and accept that some emails don't get delivered, in order to reduce complexity.

But worth revisiting this again now that we're about to merge.

Reply:
It looks like `process_notification` returns `True/False` depending if the notification was actually processed. One of the scenarios where it would return False is where the notification has had its emails "disabled". In this instance, I don't think we'd want to mark the follow up as successful / sent, would we? Since there has been no attempt to even send the email.

So I guess it's fair enough not to assume an exception means that an email hasn't been sent, but if the process_notification returns false then we should be pretty certain that an email wasn't even attempted.

Reply:
Closing as per discussion in dev meeting. In the documentation, I'll include a note about this.